### PR TITLE
Upstream sync 2021-03-31

### PR DESCRIPTION
**Pull request by pellared**
### Enable JetBrains Tools in the Benchmarks projects ([#1318](https://github.com/DataDog/dd-trace-dotnet/pull/1318))

commit cee651f3a473d353b795779d1539fb696afeb019 (datadog/master)
Author: Tony Redondo <tony.redondo@datadoghq.com>
Date:   Tue Mar 30 23:17:57 2021 +0200

* Enable JetBrains Tools

* Change comparison.

* Updates license file.

* fix license

* Update LICENSE-3rdparty.csv

Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>

Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>

### Update ASP.NET / MVC / WebApi2 Resource Names ([#1288](https://github.com/DataDog/dd-trace-dotnet/pull/1288))

commit 8aa7722a52809e0cee8fbc4cebe868297c3a1f6a
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Tue Mar 30 08:49:51 2021 +0100

* Update ASP.NET MVC span resource names

Use the route template as the resource name

* Update the ASP.NET parent span resource name to be the same as the MVC span

* Add automatic trimming of optional parameters that aren't provided

Only downside of this (other than performance), is that routes where the optional parameter is provided have different resource names compared to routes where the optional parameter is ommitted. However this is current behaviour anyway, so probably not a problem

* Add feature flag for aspnet resource names

* Propagate WebApi2 resource name to root span

The static HttpContext was the only way I could find to do this, as the ControllerContext doesn't expose it

* Update WebApi2 resource names to start with a /

* Add ASP.NET integration tests for new feature flag

* Optimization: Don't set resource name of span until after the request has completed.

No point cleaning the path in general, as in most cases it will be completely replaced

* Refactor UrlHelpers

Remove unused trivial overload of GetRelativeUrl only called by other overload
Renamed remaining method to GetCleanUriPath() as that's what it really does, and was only called with tryRemoveIds set to true
Added tests for behaviour

* Fix incorrect comments

* Remove unused parameter

* Remove cases that fail on linux (as they're testing Uri implementation, not our code)

* Fix comment

Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>

* Rename CleanUriSegment->GetCleanUriPath

* Fix feature flag

Use GetBool for resource names
Netstandard flag is special case

* Reinstate deprecated/obsolete methods to avoid issues when library reversion mismatch

* Update comments for feature flag

Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>

* Fix comment

Co-authored-by: Zach Montoya <zach.montoya@datadoghq.com>

* Move HttpContext.Items resource name key to a shared constants file

The file is linked into both places, as didn't really make sense to be in Datadog.Trace I think

* Rename Feature Flag to DD_TRACE_ROUTE_TEMPLATE_RESOURCE_NAMES_ENABLED

* Record feature flag setting in tracer debug log

Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>
Co-authored-by: Zach Montoya <zach.montoya@datadoghq.com>

### enable Service Fabric Service Remoting instrumentation out-of-the-box ([#1234](https://github.com/DataDog/dd-trace-dotnet/pull/1234))

commit cf4b751be3676d46093301d7f3204e86668ec187
Author: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>
Date:   Mon Mar 29 10:00:22 2021 -0400

### Add a consolidated multi-stage pipeline ([#1320](https://github.com/DataDog/dd-trace-dotnet/pull/1320))

commit 74d81fd710a0efb98fcc444e78deaced530dc224
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Mon Mar 29 10:30:19 2021 +0100

* Baby steps for quote-unquote ultimate pipeline. Managed unit tests first

* Use an install-dotnet pipeline to simplify dotnet installation. Test usage on the managed pools first before trying to use on arm64 unit test job

* Add install-dotnet template to build step

* Remove parameter from install-dotnet.yml

* Fix indentation in ultimate-pipeline.yml

* Add no build argument to the unit test runs

* Build the windows profiler after running the managed build

* Run the windows native unit tests on the same machine for x86 and x64

* Set buildConfiguration to Release. Remove --no-build argument on unit tests for now

* Use one Datadog.Trace.proj per build step. For build_managed, that is BuildCsharp. For unit_tests_managed, that is BuildCsharpTests.

* Add --no-build argument to dotnet pack. This means we'll build our binaries once for the BuildCsharp target and then pack the result of that

* Update publish command in build_managed job

* Add succeededOrFailed condition to native unit tests

* Fix native test that only throws after my pipeline changes

* Attempt to fix managed unit tests

* Try and get Windows build running

* Small native unit test fix

* Add the alpine and linux profiler

* Combine managed and native unit tests into same build stage

* Publish linux packages from source build

* Modify intermediate artifacts to more clearly identify some artifacts as build-only

* Attempt to add benchmarks stage with "condition: true" so that, if it works, we can disable it until it's needed

* Modify build and windows build to make it easier to build without passing custom properties everywhere

* Modify package.sh to rely on src/bin/tracer-home

* Fix benchmarks stage

* Turn off benchmark runs because they seem to run just fine

* Fix packaging build for linux containers

* Rename default tracer home directory to windows-tracer-home

* Add S3 upload

* Build ARM64 linux profiler

* Try to fix flaky unit test build...

* Build the macOS profiler in the build phase

* Actually do a git checkout in the upload step and add debugging info

* Try to make the build-linux-tracer-home-arm64 as fully fleshed out as other linux tracer homes

* Try to make the build-macos-tracer-home as fully fleshed out as other linux tracer homes

* Try to add runner tool to pipeline

* Fix tool build...hopefully

* Fix all the dotnet CLI tasks to correctly pass the build configuration

* Modify hard-coded release to instead use buildConfiguration variable

* Copy over the windows tracer home, including the .NET assemblies, into the dotnet tool

* Try windows copy again

* Last time I swear

* Get linux integration tests up and running in the unified pipeline

* Add Alpine Linux now

* Revise Linux and Alpine Linux integration tests to not checkout repo

* Attempt linux arm64 integration test run

* Add initial set of Windows integration-tests. Get rid of duplication later...

* Do explicit restore to try and fix framework reproductions

* Upload artifacts for integration-test directories to help diagnose the issue with them

* Point the managed profiler directory to Windows Tracer Home. This should work!

* Fix the TestAllPackageVersions: true to only apply to non-Windows integration test runs

* Need to install the earlier .NET Core SDKs in integration test jobs

The runtime doesn't install the asp.net crore runtime, whereas the ask does

* Extract isMainBranch variable to global

* Pass $(tracerHome) variable to MSI builder

* Add Crank stage to pipelin

* Only run pipeline on master for now

* Disable runs on PRs

* Undo changes to clr_helper_test.cpp

* Restore the publishing of managed profiler assets in the 'docker-compose run Profiler' step to fix the main integration-tests build

Co-authored-by: Zach Montoya <zach.montoya@datadoghq.com>

### Refactor all CallTarget BeginMethod methodSpec definitions to a single generic one ([#1279](https://github.com/DataDog/dd-trace-dotnet/pull/1279))

commit 4541013aa89cc5fc2c8947fb03f28a5854409726
Author: Tony Redondo <tony.redondo@datadoghq.com>
Date:   Fri Mar 26 18:14:13 2021 +0100

* Refactor all calltarget BeginMethod methodSpec definition to a single generic one.

* Fix

* Update src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

Co-authored-by: Zach Montoya <zach.montoya@datadoghq.com>

Co-authored-by: Zach Montoya <zach.montoya@datadoghq.com>

### Fix components in LICENSE-3rdparty.csv file ([#1319](https://github.com/DataDog/dd-trace-dotnet/pull/1319))

commit b12c06ba0c14ad61a1d2b4cbdf2aa573c3392c1b
Author: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>
Date:   Fri Mar 26 10:26:13 2021 -0400

### CIApp: Commit check ([#1312](https://github.com/DataDog/dd-trace-dotnet/pull/1312))

commit b851872819cf0e884f82e9c6c78a1a645a2d7ebf
Author: Tony Redondo <tony.redondo@datadoghq.com>
Date:   Thu Mar 25 21:11:31 2021 +0100

### Disable log rate limiting by default ([#1307](https://github.com/DataDog/dd-trace-dotnet/pull/1307))

commit 7e03e0f479156efb3000f7895331e56b3b4d2a0c
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Thu Mar 25 14:21:46 2021 +0000

* Disable log rate limiting by default

* Switch to using constant

### Don't set the service name in the span constructor ([#1294](https://github.com/DataDog/dd-trace-dotnet/pull/1294))

commit ccfe3c3e35503f901dfdb27e200764a26004c055
Author: Kevin Gosse <krix33@gmail.com>
Date:   Thu Mar 25 14:34:21 2021 +0100

The service name property is bound to the context, so we're setting the context service name to the value of the context service name.

### Call a local endpoint in DuplicateTypeProxy test ([#1308](https://github.com/DataDog/dd-trace-dotnet/pull/1308))

commit 536d3abf3515552103a1e8e90befd024623139e3
Author: Kevin Gosse <krix33@gmail.com>
Date:   Thu Mar 25 13:03:00 2021 +0100

* Call a local endpoint in DuplicateTypeProxy test

* Make sure the listener exits properly at the end

### Compute top-level spans on the tracer side ([#1303](https://github.com/DataDog/dd-trace-dotnet/pull/1303))

commit fd1339bc2d39aadca4251a30477a6ffa285f9e2c
Author: Kevin Gosse <krix33@gmail.com>
Date:   Thu Mar 25 11:00:05 2021 +0100

* Compute top-level spans on the tracer side

This is a requirement to implement flushing of partial traces.

### CallTarget ADO.NET performance improvements. ([#1305](https://github.com/DataDog/dd-trace-dotnet/pull/1305))

commit 676babc039e54d7b1b83d76b098531fa2dc561c8
Author: Tony Redondo <tony.redondo@datadoghq.com>
Date:   Wed Mar 24 19:43:26 2021 +0100

* Improve performance of CallTarget ADO.NET instrumentation.

* Update src/Datadog.Trace.ClrProfiler.Managed/ScopeDBFactory.cs

Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>

* Apply suggestions from code review

Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>

* Add unit tests

* Fix instrumentation from base type.

Co-authored-by: Lucas Pimentel-Ordyna <lucas.pimentel@datadoghq.com>

### Add log rotation for native logger ([#1296](https://github.com/DataDog/dd-trace-dotnet/pull/1296))

commit e51c8d2cfd396df40d363d14e61fe93cee3b0f14
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Wed Mar 24 12:40:59 2021 +0000

* Enable the rotating file sink for native logger
* Include the pid and process name in the rotating logger
The pid ensures that only one process will be rolling the file at a time. The name is for convenience.
* Strip extension from current process name in log filename

### Define benchamarks scheduled runs in yaml ([#1299](https://github.com/DataDog/dd-trace-dotnet/pull/1299))

commit e1b36525d9041cc50cbe758ace35830644bd7d56
Author: Andrew Lock <andrew.lock@datadoghq.com>
Date:   Wed Mar 24 12:38:40 2021 +0000

Currently this is defined via the "classic" editor in the UI

### CallTarget: Benchmark tests ([#1300](https://github.com/DataDog/dd-trace-dotnet/pull/1300))

commit af0cd09824345f6cc3698458ec3d5534bb6af31f
Author: Tony Redondo <tony.redondo@datadoghq.com>
Date:   Tue Mar 23 23:21:55 2021 +0100

* Initial migration of benchmarks to calltarget.

* Elastic and GraphQL and Redis benchmarks migration to calltarget.

* Fix AdoNet benchmarks

* Fix GraphQL benchmark for calltarget.

* Changes based in the review.

* Fix CI execution.

### Move constant agent headers to a central place, to avoid duplication ([#1302](https://github.com/DataDog/dd-trace-dotnet/pull/1302))

commit 2c9ff3cadf659fc8c2d3dc7d53a8f43a316d5ada
Author: Kevin Gosse <krix33@gmail.com>
Date:   Tue Mar 23 17:09:39 2021 +0100

Also move LanguageInterpreter and LanguageVersion to the constant header

### CODE REVIEW

**Comment by pellared** [comment]
```
+ cp /var/log/opentelemetry/dotnet/dotnet-tracer-native.log /project/
cp: cannot stat '/var/log/opentelemetry/dotnet/dotnet-tracer-native.log': No such file or directory
```

**Comment by pellared** [comment]
@zacharycmontoya this tests seems to be unstable:

```
[xUnit.net 00:00:11.95]     Datadog.Trace.IntegrationTests.DiagnosticListeners.AspNetCoreDiagnosticObserverTests.DiagnosticObserver_ForMvcEndpoints_SubmitsSpans(path: "/", statusCode: 200, isError: False, resourceName: "GET Home/Index", expectedTags: [[aspnet.route, ], [aspnet.controller, ], [aspnet.action, ]]) [FAIL]
  Failed Datadog.Trace.IntegrationTests.DiagnosticListeners.AspNetCoreDiagnosticObserverTests.DiagnosticObserver_ForMvcEndpoints_SubmitsSpans(path: "/", statusCode: 200, isError: False, resourceName: "GET Home/Index", expectedTags: [[aspnet.route, ], [aspnet.controller, ], [aspnet.action, ]]) [3 s]
  Error Message:
   The collection was expected to contain a single element, but it was empty.
  Stack Trace:
     at Datadog.Trace.IntegrationTests.DiagnosticListeners.AspNetCoreDiagnosticObserverTests.AssertDiagnosticObserverSubmitsSpans[T](String path, HttpStatusCode statusCode, Boolean isError, String resourceName, SerializableDictionary expectedTags) in /project/test/Datadog.Trace.IntegrationTests/DiagnosticListeners/AspNetCoreDiagnosticObserverTests.cs:line 90
   at Datadog.Trace.IntegrationTests.DiagnosticListeners.AspNetCoreDiagnosticObserverTests.DiagnosticObserver_ForMvcEndpoints_SubmitsSpans(String path, HttpStatusCode statusCode, Boolean isError, String resourceName, SerializableDictionary expectedTags) in /project/test/Datadog.Trace.IntegrationTests/DiagnosticListeners/AspNetCoreDiagnosticObserverTests.cs:line 25
--- End of stack trace from previous location ---
Results File: /project/test/Datadog.Trace.IntegrationTests/results/_8f7fec6f4222_2021-03-31_20_14_51.trx
```

**Review by RassK — APPROVED** [review]

File: src/Datadog.Trace/Configuration/ConfigurationKeys.cs [hunk 1]
```diff
@@ -361,5 +361,21 @@ internal static class Debug
...
+
+        internal static class FeatureFlags
+        {
+            /// <summary>
+            /// Feature Flag: enables updated resource names on `aspnet.request`, `aspnet-mvc.request`,
+            /// `aspnet-webapi.request`, and `aspnet_core.request` spans. Enables `aspnet_core.mvc` spans and
+            /// additional features on `aspnet_core.request` spans.
+            /// </summary>
+            /// <seealso cref="TracerSettings.RouteTemplateResourceNamesEnabled"/>
+            public const string RouteTemplateResourceNamesEnabled = "DD_TRACE_ROUTE_TEMPLATE_RESOURCE_NAMES_ENABLED";
```
Comment (pjanotti) [inline src/Datadog.Trace/Configuration/ConfigurationKeys.cs:373]: @pellared Please, update these two.

**Review by pjanotti — APPROVED** [review]
Just a minor DD_ to OTEL_ for new variables.

File: src/Datadog.Trace/Configuration/ConfigurationKeys.cs (hunk 1 above)
Comment (pellared) [reply src/Datadog.Trace/Configuration/ConfigurationKeys.cs:373]: Nice catch

Comment (pellared) [reply src/Datadog.Trace/Configuration/ConfigurationKeys.cs]: Addressed https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/99/commits/1b5650ed5252309ca011c0c61b50b799e0219a24

**Review by pjanotti — APPROVED** [review]

File: LICENSE-3rdparty.csv [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
+Datadog.Trace.ClrProfiler.Native,https://github.com/dotnet/runtime,MIT,Copyright (c) .NET Foundation and contributors. All rights reserved.
+Datadog.Trace.ClrProfiler.Native,https://github.com/Microsoft/clr-samples,MIT,Copyright (c) .NET Foundation and contributors. All rights reserved.
+Datadog.Trace.ClrProfiler.Native,https://github.com/MicrosoftArchive/clrprofiler,MIT,Copyright (c) Microsoft Corporation.  All rights reserved.
+Datadog.Trace.ClrProfiler.Native,https://github.com/nlohmann/json,MIT,Copyright (c) 2013-2018 Niels Lohmann
+Datadog.Trace.ClrProfiler.Native,https://github.com/dropbox/miniutf,MIT,"Copyright (c) 2013 Dropbox, Inc."
+Datadog.Trace.ClrProfiler.Native,https://github.com/gabime/spdlog,MIT,Copyright (c) 2016 Gabi Melman.
+Datadog.Trace.ClrProfiler.Native,https://github.com/fmtlib/fmt,MIT,"Copyright (c) 2012 - present, Victor Zverovich"
+Datadog.Trace.OpenTracing,https://github.com/opentracing/opentracing-csharp,MIT,Copyright 2016-2017 The OpenTracing Authors
+Datadog.Trace.OpenTracing,https://github.com/opentracing-contrib/csharp-netcore,Apache-2.0,
+(all),https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation,Apache-2.0,
```
Comment (zacharycmontoya) [inline LICENSE-3rdparty.csv:17]: You can remove this line for this repo 🙂

File: src/Datadog.Trace.ServiceFabric/Propagators/B3ServiceFabricContextPropagator.cs [hunk 3]
```diff
@@ -1,73 +0,0 @@
-using System;
-using System.Collections.Generic;
-using System.Text;
-using Datadog.Trace.Propagation;
-using Microsoft.ServiceFabric.Services.Remoting.V2;
-
-namespace Datadog.Trace.ServiceFabric.Propagators
-{
-    internal class B3ServiceFabricContextPropagator : IServiceFabricContextPropagator
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ServiceFabric/Propagators/B3ServiceFabricContextPropagator.cs:9]: The context propagators for Service Fabric are lost during this merge. Do you want to add it back in this PR or in a separate PR?

**Review by zacharycmontoya — COMMENTED** [review]
LGTM but I left one comment about the header propagation for Service Fabric that's getting removed in this PR

**Review by dszmigielski — APPROVED** [review]

File: src/Datadog.Trace.ServiceFabric/Propagators/B3ServiceFabricContextPropagator.cs (hunk 3 above)
Comment (pellared) [reply src/Datadog.Trace.ServiceFabric/Propagators/B3ServiceFabricContextPropagator.cs:9]: I believe that DD have refactored to use https://github.com/pellared/opentelemetry-dotnet-instrumentation/blob/upstream-2021-03-31/src/Datadog.Trace/Propagation/B3SpanContextPropagator.cs instead

@RassK Can you double-check if I am not mistaken, please?

File: LICENSE-3rdparty.csv (hunk 2 above)
Comment (pellared) [reply LICENSE-3rdparty.csv]: I changed it to `https://github.com/DataDog/dd-trace-dotnet` 😉

File: src/Datadog.Trace.ServiceFabric/Propagators/B3ServiceFabricContextPropagator.cs (hunk 3 above)
Comment (RassK) [reply src/Datadog.Trace.ServiceFabric/Propagators/B3ServiceFabricContextPropagator.cs:9]: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/99/commits/0ad53ac32195901c30a356bb3e6c80b39da9267e according to this merge, everything seems correct. The original code is already based on existing propagator, so special SF propagators must be removed.

**Comment by pellared** [comment]
Unstable test failed:
```
[xUnit.net 00:01:24.09]     Datadog.Trace.ClrProfiler.IntegrationTests.HttpMessageHandlerTests.HttpClient_SubmitsTraces(inlining: EnableCallTarget=True,EnableInlining=False, instrumentation: InstrumentSocketHandler=True,InstrumentWinHttpOrCurlHandler=False, enableSocketsHandler: False) [FAIL]
  Failed Datadog.Trace.ClrProfiler.IntegrationTests.HttpMessageHandlerTests.HttpClient_SubmitsTraces(inlining: EnableCallTarget=True,EnableInlining=False, instrumentation: InstrumentSocketHandler=True,InstrumentWinHttpOrCurlHandler=False, enableSocketsHandler: False) [2 s]
  Error Message:
   Process exited with code -532462766
Expected: True
Actual:   False
  Stack Trace:
     at Datadog.Trace.ClrProfiler.IntegrationTests.HttpMessageHandlerTests.HttpClient_SubmitsTraces(InliningOptions inlining, InstrumentationOptions instrumentation, Boolean enableSocketsHandler) in D:\a\1\s\test\Datadog.Trace.ClrProfiler.IntegrationTests\HttpMessageHandlerTests.cs:line 74
 ```

**Review by nrcventura — APPROVED** [review]

**Pull request merged**